packages/api/src/listeners/PluginServerRequest.ts

```diff
@@ -1,30 +1,62 @@
 import { Listener } from '@sapphire/framework';
 import type { ApiRequest } from '../lib/structures/api/ApiRequest';
 import type { ApiResponse } from '../lib/structures/api/ApiResponse';
-import { ServerEvents } from '../lib/structures/http/Server';
+import type { MethodName } from '../lib/structures/http/HttpMethods';
+import { ServerEvent } from '../lib/structures/http/Server';
+import { RouterRoot } from '../lib/structures/router/RouterRoot';
 
 export class PluginListener extends Listener {
 	public constructor(context: Listener.LoaderContext) {
-		super(context, { emitter: 'server', event: ServerEvents.Request });
+		super(context, { emitter: 'server', event: ServerEvent.Request });
 	}
 
 	public override async run(request: ApiRequest, response: ApiResponse) {
-		const route = this.container.server.routes.match(request);
+		const { parts, querystring } = this.#parseURL(request.url);
+		request.query = Object.fromEntries(new URLSearchParams(querystring).entries());
+
+		const branch = this.container.server.routes.router.find(parts);
+		const node = branch ? branch.node : null;
+		const route = node ? node.get((request.method ?? 'GET') as MethodName) : null;
+
+		if (node !== null) {
+			request.params = node!.extractParameters(parts);
+		}
+
+		request.routerNode = node;
+		request.route = route;
 
 		try {
 			// Middlewares need to be run regardless of the match, specially since browsers do an OPTIONS request first.
-			await this.container.server.middlewares.run(request, response, route);
+			await this.container.server.middlewares.run(request, response);
 		} catch (error) {
-			this.container.server.emit(ServerEvents.MiddlewareError, error, { request, response, route });
+			this.container.server.emit(ServerEvent.MiddlewareError, error, request, response);
 
 			// If a middleware errored, it might cause undefined behavior in the routes, so we will return early.
 			return;
 		}
 
-		if (route === null) {
-			this.container.server.emit(ServerEvents.NoMatch, request, response);
+		if (branch === null) {
+			this.container.server.emit(ServerEvent.RouterBranchNotFound, request, response);
+		} else if (route === null) {
+			this.container.server.emit(ServerEvent.RouterBranchMethodNotAllowed, request, response);
+		} else {
+			this.container.server.emit(ServerEvent.RouterFound, request, response);
+		}
+	}
+
+	#parseURL(url = '') {
```

Comment:
why is this not just using `new URL`?

Reply:
1. It's been like that since `@klasa/api` times (and `new URL()` existed back then), so changing this would be outside the scope
2. It doesn't get a full URL, but a path + querystring
3. Performance